### DISCUSSION TITLE
Should the Micronaut build plugins set a dev environment during gradle run / mn:run

### DISCUSSION THREAD

**Discussion by sdelamo** [micronaut-projects/micronaut-core#9482] [Ideas]
> The only problem with this is that the MICRONAUT_ENVIRONMENTS variable is not additive, so if the user has something defined via ApplicationContextBuilder, it will be overriden

**Comment by sdelamo** [comment]
I like the idea but I think we are late for this change in Micronaut Framework 4. I think we should consider how this change impacts [default environment](https://docs.micronaut.io/latest/guide/#_default_environment). 

I think what we could do is generate every new app in starter with : 

```
public class Application {
    @ContextConfigurer
    public static class Configurer implements ApplicationContextConfigurer {
        @Override
        public void configure(@NonNull ApplicationContextBuilder builder) {
            builder.defaultEnvironments("dev");
        }
    }
    public static void main(String[] args) {
        Micronaut.run(Application.class, args);
    }
}
```

**Comment by sdelamo** [comment]
@micronaut-projects/core-developers thoughts?

**Comment by timyates** [comment]
Adding the ApplicationContextConfigurer would always run in the dev environment though wouldn't it, unless something was done when the code was pushed to production?

One confusion I sometimes see is we have `developmentOnly` dependencies (at least for Gradle builds) which are added (eg for microstream-rest) which are unrelated to running with the Application environment containing "dev", and exist only when we do `./gradlew run`

**Comment by sdelamo** [comment]
True, Micronaut Gradle plugin has a "development" mode already. When you do gradle run, it uses the `developmentOnly` Gradle configuration.

**Comment by graemerocher** [comment]
this should be implemented in the plugins if it is to be implemented. `./gradlew run` and `mvn mn:run` should activate these. However probably too late for 4.0.x

**Comment by melix** [comment]
Having something like `builder.defaultEnvironments("dev")` sounds quite dangerous to me. Say something goes to production and no environment is set, then the dev environment would be used.

Maybe we should have a _different_ environment variable, say `MICRONAUT_DEV_ENVIRONMENTS` which would be additive once the actual set of environments has been computed by Micronaut (using plural for consistency with ENVIRONMENTS).

One of the issues with having the build plugins do magic is that this won't work unless delegating to the IDE, which is the default for Gradle + Intellij IDEA CE, but not for Maven nor IDEA Ultimate.

**Reply by sdelamo** [reply]
yes, we have to bear in mind that not everyone may want to use our build plugins.

**Reply by graemerocher** [reply]
right, IMO we should probably do something that detects some dev environment only file like `micronaut-cli.yml` or something, but doing this is not easy because the root directly changes depending whether you run from Gradle, Maven or the IDE

**Comment by sdelamo** [comment]
> Say something goes to production and no environment is set, then the dev environment would be used.

Exactly. Until now it was not as easy to happen because cloud environment detection was turned on and in most cases a cloud env would be detected and thus the default env was not being used.

**Comment by melix** [comment]
So, our build plugins are already capable of injecting dev dependencies (e.g `developmentOnly`). So we could provide a new module which exposes an `ApplicationContextConfigurer` which does this:

```
    @ContextConfigurer
    public class DevelopmentEnvironmentConfigurer implements ApplicationContextConfigurer {
        @Override
        public void configure(@NonNull ApplicationContextBuilder builder) {
            builder.defaultEnvironments("dev");
        }
    }
```

In Maven you'd have this in `provided` scope.

**Reply by graemerocher** [reply]
would this work in the IDE?

**Reply by melix** [reply]
This would work for Maven as far as I can tell. For Gradle if you delegate to Gradle for builds.

**Reply by graemerocher** [reply]
which IntelliJ ultimate doesn't do :(

**Reply by melix** [reply]
It doesn't but in any case there are many things broken if you don't delegate to Gradle. File watching will not work for example. Test resources are already broken if you don't delegate either. The `developmentOnly`  configuration is a Micronaut only thing, so Jetbrains should implement the required things if they don't want to delegate to Gradle, for whatever reason.

**Reply by sdelamo** [reply]
@melix 

Are you suggesting: 

A)  we create a module such as: `io.micronaut:micronaut-development`
B) Inside that module, we add a class with contents: 

```java
    @ContextConfigurer
    public class DevelopmentEnvironmentConfigurer implements ApplicationContextConfigurer {
        @Override
        public void configure(@NonNull ApplicationContextBuilder builder) {
            builder.defaultEnvironments("dev");
        }
    }
}
```

C) We add the dependency transparently with `developmentOnly("io.mcronaut:mcironatu-development") ` through the Gradle Plugin. 

The result of that would be to run with `dev` environment when running `gradle run`. Moreover,  the default environment `dev` will not leak to production since a `developmentOnly` dependency will not be leaked into the production classpath. 

I like the idea. 

However, What will happen if a user set a default environment directly in their app?

**Reply by melix** [reply]
Yes, that's what I'm suggesting.

> However, What will happen if a user set a default environment directly in their app?

as far as I can tell, the default environment would be applied, `builder.defaultEnvironments` is additive.
